node/src/types/block.rs

```diff
@@ -747,21 +759,60 @@ impl BlockHeader {
 
     /// Hash of the block header.
     pub fn hash(&self) -> BlockHash {
-        let serialized_header = Self::serialize(&self)
-            .unwrap_or_else(|error| panic!("should serialize block header: {}", error));
-        BlockHash::new(hash::hash(&serialized_header))
+        // Pattern match here leverages compiler to ensure every field is accounted for
+        let BlockHeader {
+            parent_hash,
+            era_id,
+            body_hash,
+            state_root_hash,
+            era_end,
+            height,
+            timestamp,
+            protocol_version,
+            random_bit,
+            accumulated_seed,
+        } = self;
+
+        let hashed_era_end = match era_end {
+            None => hash::hash(&[]),
```

Comment:
(We had talked about maybe using all zeros instead of `hash(&[])`. In this case it doesn't matter, though, because `era_end.hash()` probably can't be `hash(&[])`.)

Reply:
FWIW, in other places where we have this kind of "zero element hash" we hash all zeros array.

Maybe, we can have a separate module for these new functions and add a constant `const EMPTY_HASH = hash::hash([0u8; 32])` so that people are not tempted to use them in different contexts? This pattern seems to be used throughout the EE codebase as well where the functions are grouped together and there are constants that hide internal details/design decisions.